web-client/src/views/FileDocument/StateDrivenFileInput.tsx

```diff
@@ -48,6 +57,53 @@ export const StateDrivenFileInput = connect<
 
     const fileOnForm = file || form[fileInputName] || form.existingFileName;
 
+    // Setting the filename here so that we can display it before validation
+    // finishes, otherwise a slow machine might have slight lag and allow the user
+    // to "choose file" again.
+    const [selectedFilename, setSelectedFilename] = useState('');
+
+    const onFileSelectionChange = async (
```

Comment:
This is not a very "Cerebral" way to do things, which would involve using a sequence (although we would still need to do a little DOM manipulation to unset `e.target.value` on error.)

I have heard mixed opinions about making this more Cerebral-esque. I spent a few hours going that direction to see what it would look like. Right now, my overall feeling is that the advantages of moving things into Cerebral in this case (consistency, extensibility of sequences, doing more outside of the component rather than within it) do not necessarily outweigh the disadvantages (opacity of types, potential over-abstraction, slightly more time to finish). If enough people strongly disagree, I can switch to my Cerebral-work branch and continue that direction.

Note that I think this entire file (and `ScanBatchPreview.tsx` and `PetitionQxScanBatchPreview.tsx`) should be refactored to be more agnostic about the state library. (E.g., this file should probably just have a function passed in for onChange.) I will open a DevEx ticket for this.